alf/networks/transformer_networks.py

```diff
@@ -0,0 +1,175 @@
+
+        for i in range(num_prememory_layers):
+            core, inputs = T_i([core, inputs], [core, inputs])
+        for j in range(num_memory_layers):
+            core, inputs = TM_j([memory_j, core, inputs], [core, inputs])
+            memory_j.write(core)
+        return core, new_memory_state
+
+    where T_i denotes the ``TransformerBlock``  for the i-th prememory layers
+    and TM_j denotes the ``TransformerBlock`` for the j-th memory layers.
+    """
+
+    def __init__(self,
+                 input_tensor_spec,
+                 memory_size,
+                 core_size,
+                 num_prememory_layers,
+                 num_memory_layers,
+                 num_attention_heads,
+                 d_ff,
+                 centralized_memory=True,
+                 input_preprocessors=None,
+                 name="TransformerNetwork"):
+        """
+        Args:
+            input_tensor_spec (nested TensorSpec): the (nested) tensor spec of
+                the input.
+            memory_size (int): size of memory.
+            core_size (int): size of core (i.e. number of embeddings of core)
+            num_prememory_layers (int): number of TransformerBlock calculation
+                without using memory
+            num_memory_layers (int): number of TransformerBlock calculation
+                using memory
+            num_attention_heads (int): number of attention heads for each
+                ``TransformerBlock``
+            d_ff (int): the size of the hidden layer of the feedforward network
+                in each ``TransformerBlock``
+            centralized_memory (bool): if False, there will be a separate memory
+                for each memory layers. if True, there will be a single memory
+                for all the memroy layers and it is updated using the last core
+                embeddings.
+            input_preprocessors (nested Network|nn.Module): a nest of
+                preprocessor networks, each of which will be applied to the
+                corresponding input. If not None, then it must have the same
+                structure with ``input_tensor_spec``. If any element is None, then
+                it will be treated as math_ops.identity. This arg is helpful if
+                you want to have separate preprocessings for different inputs by
+                configuring a gin file without changing the code. For example,
+                embedding a discrete input before concatenating it to another
+                continuous vector. The output_spec of each input preprocessor i
+                should be [input_size_i, d_model]. The result of all the preprocessors
+                will be concatenated as a Tensor of shape ``[batch_size, input_size, d_model]``,
+                where ``input_size = sum_i input_size_i``.
+
+        """
+        super().__init__(
+            input_tensor_spec,
+            input_preprocessors,
+            preprocessing_combiner=NestConcat(dim=-2),
```

Comment:
Why it's -2? The input_tensor_spec argument doesn't say anything about the input format.